mmdet/evaluation/metrics/coco_panoptic_metric.py

```diff
@@ -226,66 +236,181 @@ def result2json(self, results: Sequence[dict],
+    def _parse_predictions(self,
+                           pred: dict,
+                           img_id: int,
+                           segm_file: str,
+                           label2cat=None) -> dict:
+        """Parse panoptic segmentation predictions.
 
         Args:
-            data_batch (Sequence[dict]): A batch of data
-                from the dataloader.
-            predictions (Sequence[dict]): A batch of outputs from
-                the model.
+            pred (dict): Panoptic segmentation predictions.
+            img_id (int): Image id.
+            segm_file (str): Segmentation file name.
+            label2cat (dict): Mapping from label to category id.
+                Defaults to None.
+
+        Returns:
+            dict: Parsed predictions.
         """
-        os.makedirs(self.seg_out_dir, exist_ok=True)
+        result = dict()
+        result['img_id'] = img_id
+        # shape (1, H, W) -> (H, W)
+        pan = pred['pred_panoptic_seg']['sem_seg'].cpu().numpy()[0]
+        pan_labels = np.unique(pan)
+        segments_info = []
+        for pan_label in pan_labels:
+            sem_label = pan_label % INSTANCE_OFFSET
+            # We reserve the length of self.CLASSES for VOID label
+            if sem_label == len(self.dataset_meta['CLASSES']):
+                continue
+            mask = pan == pan_label
+            area = mask.sum()
+            segments_info.append({
+                'id':
+                int(pan_label),
+                # when ann_file provided, sem_label should be cat_id, otherwise
+                # sem_label should be a continuous id, not the cat_id
+                # defined in dataset
+                'category_id':
+                label2cat[sem_label] if label2cat else sem_label,
+                'area':
+                int(area)
+            })
+        # evaluation script uses 0 for VOID label.
+        pan[pan % INSTANCE_OFFSET == len(self.dataset_meta['CLASSES'])] = VOID
+        pan = id2rgb(pan).astype(np.uint8)
+        mmcv.imwrite(pan[:, :, ::-1], osp.join(self.seg_out_dir, segm_file))
+        result = {
+            'image_id': img_id,
+            'segments_info': segments_info,
+            'file_name': segm_file
+        }
+
+        return result
+
+    def process_with_tmpfile_without_json(self, data_batch: Sequence[dict],
+                                          predictions: Sequence[dict]) -> None:
+        """Process gts and predictions when ``outfile_prefix`` and ``ann_file``
+        are not set.
+
+        In this case, gts are from dataset. And the ``pq_stats`` is computed
+        here.
+        """
+        categories = dict()
+        for id, name in enumerate(self.dataset_meta['CLASSES']):
+            isthing = 1 if name in self.dataset_meta['THING_CLASSES'] else 0
+            categories[id] = {'id': id, 'name': name, 'isthing': isthing}
 
         for data, pred in zip(data_batch, predictions):
```

Comment:
Suggest refactoring this method to process only one image. Because after mmengine refactoring dataflow, the databatch may no longer be a list.